nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/disks.rs

```diff
@@ -104,18 +148,63 @@ impl DisksEditor {
         let did_expunge: bool;
         match config.disposition {
             BlueprintPhysicalDiskDisposition::InService => {
-                config.disposition = BlueprintPhysicalDiskDisposition::Expunged;
+                config.disposition =
+                    BlueprintPhysicalDiskDisposition::Expunged {
+                        // We don't update the editor generation until the call
+                        // to `finalize` which occurs later. We bump it here
+                        // to compensate.
+                        as_of_generation: self.generation.next(),
+                        ready_for_cleanup: false,
+                    };
                 self.counts.expunged += 1;
                 did_expunge = true;
             }
-            BlueprintPhysicalDiskDisposition::Expunged => {
+            BlueprintPhysicalDiskDisposition::Expunged { .. } => {
                 // expunge is idempotent; do nothing
                 did_expunge = false;
             }
         }
 
         Ok((did_expunge, config.pool_id))
     }
+
+    pub fn decommission(
+        &mut self,
+        disk_id: &PhysicalDiskUuid,
+    ) -> Result<bool, DisksEditError> {
+        let config = self.disks.get_mut(disk_id).ok_or_else(|| {
+            DisksEditError::DecommissionNonexistentDisk { id: *disk_id }
+        })?;
+
+        match config.disposition {
+            BlueprintPhysicalDiskDisposition::InService => {
+                return Err(DisksEditError::DecommissionInServiceDisk {
+                    id: *disk_id,
+                });
+            }
+            BlueprintPhysicalDiskDisposition::Expunged {
+                ready_for_cleanup: false,
+                as_of_generation,
+            } => {
+                config.disposition =
+                    BlueprintPhysicalDiskDisposition::Expunged {
+                        ready_for_cleanup: true,
+                        as_of_generation,
+                    };
+                self.counts.decommissioned += 1;
```

Comment:
Hmm, will changing `counts` here cause us to bump the generation on the `PhysicalDisksConfig` we emit? That doesn't seem right. Could we return `true` here without modifying `counts`? (At that point maybe `counts` doesn't need the new `decommissioned` field?)

Reply:
Good call, this is what we just discussed wrt your upcoming changes. I added a relevant comment in a test for this as well. Done in 55a65f8d01ed8e54a7ade5b84eeae83a937956dd